### PR TITLE
Hash my url

### LINKED ISSUE #1474: Dashboard: Single use booking link
Allow people to create disposable booking links, which don't give away their regular booking links.

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

**Adds**:
- a new path `/d/[link]` for temporary links (can be used for disposable booking links when that feature is implemented)
- a new booking page `/d/book` for booking using temporary links (can be used for disposable booking links when that feature is implemented)
- adds a new model for hashed links/urls
```
model HashedLink {
  id          Int       @id @default(autoincrement())
  link        String    @unique()
  eventType   EventType @relation(fields: [eventTypeId], references: [id], onDelete: Cascade)
  eventTypeId Int       @unique
}
```

**Feature**:
- Users can enable a setting in the event-type which creates a hash URL they can share with the clients they don't want to reveal their usernames/slugs
- Once the booking is complete using this link, the link automatically expires and a new one is generated in its place
- The regular event-type link works normally, this hashed URL is usable as a separate feature for the same event-type

**Demo**:
[Loom Video](https://www.loom.com/share/0a91d90784344a0184009097d3a485a3)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Tried pushing changes to `staging` (app.cal.dev) but deployment failed due to unrelated error.

Can test on staging with the specific preview deployment with just a little bit of hassle:
[Loom Here](https://www.loom.com/share/198da89aefca498895b5726f68173a94)

Can also be tested locally by pulling the branch. Follow the steps in loom [here](https://www.loom.com/share/0a91d90784344a0184009097d3a485a3)

Playwright tests have been added to confirm that 
- [x] Hashed URL is being created successfully
- [x] Hashed URL is being used to book successfully
- [x] Upon booking, a new hashed URL is being generated which replaces the previous one

## Checklist

- This PR needs changes to the documentation

### CODE REVIEW

File: apps/web/pages/[user]/book.tsx [hunk 1]
```diff
@@ -196,6 +196,8 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
       booking,
       trpcState: ssr.dehydrate(),
       isDynamicGroupBooking,
+      hasHashedBookingLink: false,
+      hashedLink: undefined,
```
Comment (alishaz-polymath) [inline apps/web/pages/[user]/book.tsx:200]: To confirm the possible associated hashed link shouldn't be updated after the event is booked

File: apps/web/pages/api/book/event.ts [hunk 2]
```diff
@@ -772,6 +773,20 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+  if (hasHashedBookingLink) {
+    await prisma.hashedLink.update({
+      where: {
+        link: reqBody.hashedLink as string,
+      },
+      data: {
+        link: hashedUid,
+      },
+    });
+  }
```
Comment (alishaz-polymath) [inline apps/web/pages/api/book/event.ts:789]: Hashed link is updated if used for booking

File: apps/web/pages/event-types/[type].tsx [hunk 3]
```diff
@@ -442,6 +445,10 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
     team ? `team/${team.slug}` : eventType.users[0].username
   }/${eventType.slug}`;
 
+  const placeholderHashedLink = `${process.env.NEXT_PUBLIC_WEBSITE_URL}/d/${
+    eventType.hashedLink ? eventType.hashedLink.link : "xxxxxxxxxxxxxxxxx"
+  }/${eventType.slug}`;
```
Comment (alishaz-polymath) [inline apps/web/pages/event-types/[type].tsx:450]: The placeholder hashed link to show in the input field and for copying purposes, if the hashed link is already generated, else a fallback placeholder which won't be copied until saved/updated

File: apps/web/pages/event-types/[type].tsx [hunk 4]
```diff
@@ -1330,6 +1338,65 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
...
+                                          <DocumentDuplicateIcon className="w-6 p-1 text-neutral-500" />
+                                        </Button>
+                                      </Tooltip>
+                                    </div>
+                                    <span className="text-xs text-gray-500">
+                                      The URL will regenerate after each use
+                                    </span>
+                                  </div>
+                                </div>
+                              )}
```
Comment (alishaz-polymath) [inline apps/web/pages/event-types/[type].tsx:1395]: Show hashed link placeholder if the "Generate hash link" field is checked

File: apps/web/server/routers/viewer/eventTypes.tsx [hunk 5]
```diff
@@ -234,6 +246,48 @@ export const eventTypesRouter = createProtectedRouter()
...
+              link: uid,
+            },
+            create: {
+              link: uid,
+              eventType: {
+                connect: { id: input.id },
+              },
+            },
+          });
+        }
```
Comment (alishaz-polymath) [inline apps/web/server/routers/viewer/eventTypes.tsx:279]: Create a new link if there was no prior connected link. This is done if the "Generate hash link" was checked in event type update.

File: apps/web/server/routers/viewer/eventTypes.tsx [hunk 6]
```diff
@@ -234,6 +246,48 @@ export const eventTypesRouter = createProtectedRouter()
...
+      } else {
+        // check if hashed connection exists. If it does, disconnect
+        if (connectedLink) {
+          await ctx.prisma.hashedLink.delete({
+            where: {
+              eventTypeId: input.id,
+            },
+          });
+        }
```
Comment (alishaz-polymath) [inline apps/web/server/routers/viewer/eventTypes.tsx:288]: Disconnect any connected links if a connection exists and the "Generate Hashed link" field was checked in the event type update.

**Review by alishaz-polymath — COMMENTED** [review]
Self review added

**Comment by hariombalhara** [comment]
I noticed that one disposable link is shown at a time to the user. If a user shares it with 2 persons, only one of them can successfully book the meeting.
Shouldn't we allow to generate multiple links simulatenously to solve this problem ?

**Comment by alishaz-polymath** [comment]
> I noticed that one disposable link is shown at a time to the user. If a user shares it with 2 persons, only one of them can successfully book the meeting. Shouldn't we allow to generate multiple links simulatenously to solve this problem ?

That is by design. The use-case for multiple bookings with the same link will be taken care of in the "disposable booking links" feature, which solves for a different use case. The use-case in this feature is specifically meant for a one-off single use booking where the cal username is not shared with the clients

**Comment by zomars** [comment]
Minor nitpick: We keep calling it "disposable link" but we named it "hashed link" in codee. Shouldn't we just name it `DisposableLink` in the code?

**Comment by alishaz-polymath** [comment]
> Minor nitpick: We keep calling it "disposable link" but we named it "hashed link" in code. Shouldn't we just name it `DisposableLink` in the code?

After a lot of discussions between @Jaibles and I, and @Jaibles , @PeerRich and @baileypumfleet  it was concluded that we'd reserve disposable link for a follow up feature which solves for the issue [Here](https://github.com/calcom/cal.com/issues/1474)

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 7]
```diff
@@ -12,6 +12,7 @@ import { useSession } from "next-auth/react";
 import dynamic from "next/dynamic";
 import Head from "next/head";
 import { useRouter } from "next/router";
+import { HashLinkPageProps } from "pages/d/[link]/book";
```
Comment (zomars) [inline apps/web/components/booking/pages/BookingPage.tsx:15]: Do we mean to use an absolute import here?

**Review by zomars — CHANGES_REQUESTED** [review]
Code looks good! Since this is a new feature I would require two things:
1. Get this working on staging
2. Add E2E to make sure this feature works

File: apps/web/components/booking/pages/BookingPage.tsx (hunk 7 above)
Comment (alishaz-polymath) [reply apps/web/components/booking/pages/BookingPage.tsx]: Oh, right. Updating this to be similar to `BookPageProps` & `TeamBookingPageProps`

**Comment by zomars** [comment]
> 1. This works well on staging but on the specific deployment ("https://cal-q7o6xb2oa-cal.vercel.app", for example). I think this doesn't work on `dev.cal.com` yet because of migrations?

I think you mean `app.cal.dev`. Both Vercel preview deployments and this share the same DB. So if one has a migration means the other one as well.

> 2. For E2E test, the above issue might result in failed test, right? 

Not really since tests don't run on Vercel and are ran on a fresh DB.

> We would probably need the migration to be set in staging before the E2E test can actually run fine because it would expect to see this feature functional in staging, right?

Nope as explained above. E2E tests are run on a fresh environment.

**Comment by alishaz-polymath** [comment]
> > 1. This works well on staging but on the specific deployment ("https://cal-q7o6xb2oa-cal.vercel.app", for example). I think this doesn't work on `dev.cal.com` yet because of migrations?
> 
> I think you mean `app.cal.dev`. Both Vercel preview deployments and this share the same DB. So if one has a migration means the other one as well.
> 

@zomars

So, That's weird because I see the changes here (notice the `Generate hashed URL` checkbox):
image

But not here:
image

Is there a reason why this might be? Or do I have to manually push a particular branch into dev deployment?

**Comment by zomars** [comment]
You need to push to the `staging` branch.

**Comment by alishaz-polymath** [comment]
@zomars I have added the 

> Code looks good! Since this is a new feature I would require two things:
> 
>     1. Get this working on staging
> 
>     2. Add E2E to make sure this feature works

1. I pushed changes to staging but deployment fails (due to unrelated issue) and hence it can't be checked on `app.cal.dev` however, it works like a charm on preview deployments in staging. [Here's a loom](https://www.loom.com/share/198da89aefca498895b5726f68173a94) of how you can test it on preview deployments as well

2. Added E2E tests and they work

**Review by zomars — APPROVED** [review]
Functionally it works like a charm. I'm still not 100% in the UX side of things but I'll let @Jaibles be the judge of that and we can follow up later. Thanks @alishaz-polymath 🙏🏽

**Pull request merged**